<a href="https://colab.research.google.com/github/Leo-alves2/Corretor_Ortografico/blob/main/Corretor.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

O que é o processamento de linguagem natural (NLP).

O que está por trás de um corretor ortográfico.

Como ler um arquivo de texto utilizando python.

Como contar a quantidade de caracteres de um texto usando len().

In [1]:
# Ler um arquivo texto

open('artigos.txt', 'r')

<_io.TextIOWrapper name='artigos.txt' mode='r' encoding='UTF-8'>

In [2]:
# Colocando o artigo em uma variavel para leitura

with open('artigos.txt', 'r') as f:
  artigos = f.read()

print(artigos[:500])




imagem 

Temos a seguinte classe que representa um usuário no nosso sistema:

java

Para salvar um novo usuário, várias validações são feitas, como por exemplo: Ver se o nome só contém letras, [**o CPF só números**] e ver se o usuário possui no mínimo 18 anos. Veja o método que faz essa validação:

java 

Suponha agora que eu tenha outra classe, a classe `Produto`, que contém um atributo nome e eu quero fazer a mesma validação que fiz para o nome do usuário: Ver se só contém letras. E aí? Vou


# Tokenização

In [3]:
# Verificando as palavras repetidas

len(artigos)

2605046

In [4]:
# Velificando uma palavra expecifica

len('Olá')

3

In [5]:
# Utilizando o split para separar palavras, não é recomendado, segue exemplo

# Criando uma frase

texto_exemplo = 'Olá, tudo bem?'

# Usado o Split para separar

palavras_separadas = texto_exemplo.split()

# Printando a variável com as palavras separadas

# Verifique que o split entende como palavra a pontuação. Ex. "bem?"

print(palavras_separadas)

['Olá,', 'tudo', 'bem?']


In [6]:
# ['Olá,', 'tudo', 'bem?'] cada elemento desse lista 
# é um Token em Processamento de Linguagem Natural (NLP).

texto_exemplo = 'Olá, tudo bem?'
tokens = texto_exemplo.split()
print(tokens)

['Olá,', 'tudo', 'bem?']


Token é uma sequência de caracteres, separados por um limitador, que pode ser um espaço em branco, pontuação ou quebra de linhas, por exemplo.

Os tokens não necessariamente são apenas palavras, podem ser uma sequência de caracteres alfanuméricos, alfabéticos com pontuações, etc.

O processo de transformar um arquivo de texto em pequenos tokens é chamado de Tokenização, o qual é recorrente no pré-processamento ou na análise estatística de dados textuais. Então, como é bastante utilizado, temos ferramentas que facilitam a sua implementação.

Uma bem conhecida na área de NLP é o nltk ou Natural Language Tool Kit, que é um conjunto de ferramentas que implementa diversos métodos e algoritmos para análise textual.

Apenas relembrando nosso problema, separamos os tokens da nossa frase com split(), mas as palavras ainda estão junto com as pontuações. Então separaremos em tokens de palavras e tokens de pontuação.

In [7]:
import nltk 
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [8]:
palavras_separadas = nltk.tokenize.word_tokenize(texto_exemplo)

In [9]:
print(palavras_separadas)

['Olá', ',', 'tudo', 'bem', '?']


In [10]:
len(palavras_separadas)

5

Para descobrirmos se um Token é palavra ou pontuação usaremos um método chamado .isalpha() se for apenas palavra retornará como verdadeiro. Caso encontre um que não seja alfabético como exemplo "palavra1" ou apenas "./", "?" com um número ou pontuação por exemplo, nos retornará falso.

In [11]:
# Criando uma função para pegar somente as palavras sem a pontuação

# Cirando um lista vazia

def separa_palavras(lista_tokens):
  lista_palavras = []
  for token in lista_tokens:
    if token.isalpha():
      lista_palavras.append(token)
  return lista_palavras

In [12]:
# Testando, teremos separado a lista de tokens somente em palavras.

separa_palavras(palavras_separadas)

['Olá', 'tudo', 'bem']

In [13]:
# Aplicando a função no arquivo

lista_tokens = nltk.tokenize.word_tokenize(artigos)
lista_palavras = separa_palavras(lista_tokens)
print(f"O número de palavras é {len(lista_palavras)}")

O número de palavras é 403031


# Normalização

O valor total de 393916 palavras retornado anteriormente não representa a quantidade de palavras únicas que podem ser corrigidas, afinal muitas delas aparecerão *repetidas* nos textos.

Portanto, precisaremos calcular quantos vocábulos únicos existem sem repetição.

In [14]:
# Printando as 5 primeiras palavras

print(lista_palavras[:5])

['imagem', 'Temos', 'a', 'seguinte', 'classe']


Neste caso, a palavra “Temos” inicia com letra maiúscula, mas também poderemos ter a mesma palavra “temos” iniciando com letra minúscula no meio do texto, então as contabilizaremos como dois tipos diferentes.

Portanto, deveremos normalizar o texto e transformar todo o texto em letras minúsculas.

In [15]:
def normalizacao(lista_palavras):
    lista_normalizada = []
    for palavra in lista_palavras:
        lista_normalizada.append(palavra.lower())
    return lista_normalizada


In [16]:
# Visualizar

# Como retorno, receberemos a mesma lista de palavras anterior, mas a palavra 
# 'temos' aparecerá com letra minúscula ao invés de maiúscula como estava antes.

lista_normalizada = normalizacao(lista_palavras)
print(lista_normalizada[:5])

['imagem', 'temos', 'a', 'seguinte', 'classe']


Agora que já temos nossa lista normalizada, poderemos remover as repetições de palavras, e existem algumas formas de fazer isso.

Poderemos tanto construir um algoritmo, quanto trabalhar com ideias mais interessantes no Python, pois este tem métodos que implementam a ideia de conjuntos matemáticos, como o set().

In [17]:
# Exemplo com SET

set([1, 2, 3, 3, 3, 3, 4, 5, 6])

{1, 2, 3, 4, 5, 6}

In [18]:
# Utilizando o Set para remover as palavras repetidas da lista_palavras

set(lista_normalizada)

{'mergulhar',
 'passaram',
 'perl',
 'oficializou',
 'avançadas',
 'lambda',
 'fechar',
 'concorrente',
 'treinamos',
 'cenário',
 'alterar',
 'bombando',
 'semente',
 'executasse',
 'missão',
 'chamativa',
 'saudate',
 'leigo',
 'map',
 'maker',
 'cabeçalhos',
 'ofício',
 'in',
 'incalculável',
 'df',
 'slugs',
 'apresentada',
 'uol',
 'encaixam',
 'datalist',
 'permitidas',
 'tcc',
 'benéfica',
 'zoeira',
 'registre',
 'marmitas',
 'tomadas',
 'sons',
 'dirigindo',
 'satisfatório',
 'atuará',
 'retrovisores',
 'prevenção',
 'usual',
 'píxels',
 'concreto',
 'designernatan',
 'matrícula',
 'leucemiazero',
 'sobrou',
 'motor',
 'sorted',
 'teclas',
 'alura',
 'disco',
 'gateway',
 'xt',
 'remotos',
 'trazido',
 'cm',
 'selecionado',
 'alemanha',
 'quão',
 'texto',
 'modelagem',
 'doadores',
 'representamos',
 'retornava',
 'princípio',
 'adaptei',
 'sólidos',
 'stanford',
 'dependemos',
 'parciais',
 'implementados',
 'módulo',
 'leads',
 'aumentaram',
 'familiarizar',
 'mantêm',
 'con

In [19]:
# Contando a quantidade de palavras

len(set(lista_normalizada))

18464

Executando o código, obteremos a quantidade de 18.464 de tipos diferentes, o que é um valor bastante interessante.

Este será o número de palavras que nosso corretor ortográfico aprenderá a corrigir, lembrando que temos um total de 403.031 vocábulos no texto.

# Detalhando o corretor

Relembrando como nosso corretor ortográfico funcionará, queremos que este receba o termo "lgica" escrito erroneamente e que devolva-o corrigido para "lógica".

Descobrimos a palavra correta porque está entre os mais de 17.000 vocábulos conhecidos pela contagem feita anteriormente.

Porém, ainda não implementamos a transição de “lgica” para “lógica” de fato.

Pegaremos a palavra digitada errada e geraremos novas palavras, onde uma delas poderá ser a correta. Neste nosso problema específico, temos a letra "ó" a menos.

Nosso modelo de NLP precisará de um algoritmo capaz de inserir uma letra a mais neste termo digitado equivocadamente.

No Processamento de Linguagem Natural, temos a entrada da palavra equivocada e a saída da correta. Entre esses pontos, elaboramos uma tabela com três colunas representando os passos da correção deste algoritmo que desenvolveremos.

<table><thead><tr><th>ESQUERDO</th><th>DIREITO</th><th>OPERAÇÃO</th><th>RESULTADO</th></tr></thead><tbody><tr><td>null</td><td>lgica</td><td>null + ó + lgica</td><td>óligca</td></tr><tr><td>l</td><td>gica</td><td>l + ó + gica</td><td><strong>lógica</strong></td></tr><tr><td>lg</td><td>ica</td><td>lg + ó + ica</td><td>lgóica</td></tr><tr><td>lgi</td><td>ca</td><td>lgi + ó + ca</td><td>lgióca</td></tr><tr><td>lgic</td><td>a</td><td>lgic + ó + a</td><td>lgicóa</td></tr><tr><td>lgica</td><td>null</td><td>lgica + ó + null</td><td>lgicaó</td></tr></tbody></table>

Neste passo, transformaremos o algoritmo no formato de tabela em código.

Como resultado das operações, geraremos novas palavras candidatas à correção. Então começaremos definindo a função gerador_palavras().

Então construiremos um algoritmo que fatiará a palavra. Em Python, poderemos utilizar o mesmo conceito de Fatiamento de Listas para strings.

Para acessarmos um elemento, chamaremos a lista[] na linha seguinte e abriremos colchetes para passarmos o index da posição que queremos. Neste caso, acessaremos o primeiro item da lista passando apenas o valor 0.

In [20]:
lista = [1,2,3]
lista[0]

1

Se quisermos o último dos três elementos, o index é 2. Com isso, além de acessarmos, poderemos fatiar a lista pegando do primeiro item ao último.

Se precisarmos retornar o elemento de posição 1 até 2 com : entre os valores de lista[], receberemos 2 como resultado, e não 3 como o esperado.

Isso aconteceu porque o sistema sempre retornará a posição anterior, ou seja, pegará até a posição 1 ao invés da 2. Então, para retornarmos o último elemento, passaremos um número a mais, neste caso 3.

In [21]:
lista = [1,2,3]
lista[1:3]

[2, 3]

Com isso, teremos fatiado a lista entre os elementos 2 e 3 como queríamos.

Faremos algo similar com a string. Passaremos "lgica" para a variável lista ao invés de [1,2,3], e pegaremos as posições 1 relativa ao "l" até a 2 que é o "i" dentro de lista[].

In [22]:
lista = "lgica"
lista[1:3]

'gi'

Como retorno, receberemos a parte 'gi' da palavra.

Portanto, trabalhamos com Fatiamento de strings da mesma maneira feita com listas, o que nos ajudará bastante.

Para entendermos melhor como fatiaremos as palavras, precisaremos pegar o lado direito e esquerdo por meio da criação de uma tupla, colocando a lista[] entre parênteses.

O primeiro elemento da tupla será representado pela primeira parte do lado esquerdo e o segundo pela segunda fatia.

In [23]:
lista = "lgica"
(lista[:1],lista[1:])

('l', 'gica')

O retorno com o lado esquerdo da primeira posição é o 'l' e do lado direito é o 'gica'. Em seguida, poderemos apenas variar os números das partes da tupla.

In [24]:
palavra_exemplo = "lgica"
def gerador_palavras(palavra):
    fatias = []
    for i in range(len(palavra)):
        fatias.append((lista[:i],lista[i:]))
    print(fatias)
    #palavras_geradas = insere_letras(fatias)
    #return palavras_geradas

gerador_palavras(palavra_exemplo)

[('', 'lgica'), ('l', 'gica'), ('lg', 'ica'), ('lgi', 'ca'), ('lgic', 'a')]


No exemplo de Fatiamento que fizemos no início, vimos que é preciso adicionar um número a mais para pegarmos o último caractere da string. Da mesma forma, adicionaremos +1 ao len() da palavra dentro de range().

In [25]:
palavra_exemplo = "lgica"
def gerador_palavras(palavra):
    fatias = []
    for i in range(len(palavra)+1):
        fatias.append((lista[:i],lista[i:]))
    print(fatias)
    #palavras_geradas = insere_letras(fatias)
    #return palavras_geradas

gerador_palavras(palavra_exemplo)

[('', 'lgica'), ('l', 'gica'), ('lg', 'ica'), ('lgi', 'ca'), ('lgic', 'a'), ('lgica', '')]


Com isso, receberemos os mesmos seis resultados de palavras fatiadas que podem ser vistos na tabela, sendo o último ('lgica','') como esperado.

Continuando, a função insere_letras() receberá as fatias da palavra e retornará as novas combinações.

# Inserir caracteres

Em seu corpo, criaremos a variável novas_palavras sendo igual a uma lista vazia [].

Dentro, inseriremos a operação que criará a palavra; novas_palavras com .append() receberá E representando o lado esquerdo, somado com a letra a ser inserida entre as partes, mais o D do lado direito. Porém, ainda não temos nenhum desses elementos definidos.

Primeiramente, definiremos as letras sendo igual a todas as letras do alfabeto e as vogais acentuadas juntas, e deveremos passar por cada caractere durante a operação.

Para isso, faremos um forde letra em letras, e realizaremos a operação dentro deste laço da variável novas_palavras.

Para definirmos o E e o D que representam cada parte da palavra, adicionaremos mais um for pegando os lados E, D em fatias, as quais retornam as tuplas.

Para cada um dos lados, realizaremos o for de letra feito antes.

In [26]:
def insere_letras(fatias):
    novas_palavras = []
    letras = 'abcdefghijklmnopqrstuvwxyzàáâãèéêìíîòóôõùúûç'
    for E, D in fatias:
        for letra in letras:
            novas_palavras.append(E + letra + D)
    return novas_palavras

Organizaremos os trechos de código em uma única célula para executarmos todos os comandos de uma vez.

In [27]:
palavra_exemplo = "lgica"

def insere_letras(fatias):
    novas_palavras = []
    letras = 'abcdefghijklmnopqrstuvwxyzàáâãèéêìíîòóôõùúûç'
    for E, D in fatias:
        for letra in letras:
            novas_palavras.append(E + letra + D)
        return novas_palavras

def gerador_palavras(palavra):
    fatias = []
    for i in range(len(palavra)+1):
        fatias.append((palavra[:i],palavra[i:]))
    palavras_geradas = insere_letras(fatias)
    return palavras_geradas

palavras_geradas = gerador_palavras(palavra_exemplo)
print(palavras_geradas)

['algica', 'blgica', 'clgica', 'dlgica', 'elgica', 'flgica', 'glgica', 'hlgica', 'ilgica', 'jlgica', 'klgica', 'llgica', 'mlgica', 'nlgica', 'olgica', 'plgica', 'qlgica', 'rlgica', 'slgica', 'tlgica', 'ulgica', 'vlgica', 'wlgica', 'xlgica', 'ylgica', 'zlgica', 'àlgica', 'álgica', 'âlgica', 'ãlgica', 'èlgica', 'élgica', 'êlgica', 'ìlgica', 'ílgica', 'îlgica', 'òlgica', 'ólgica', 'ôlgica', 'õlgica', 'ùlgica', 'úlgica', 'ûlgica', 'çlgica']


Com isso, geramos um conjunto com todas as possibilidades de correção, ou seja, a palavra separada em duas partes com inserção de letras no meio, formando novas palavras.

De volta ao nosso código, faremos uma pequena correção na função gerador_palavras(). Dentro do for, substituiremos o termo lista por palavra na tupla dentro de append().

In [28]:
palavra_exemplo = "lgica"

def insere_letras(fatias):
    novas_palavras = []
    letras = 'abcdefghijklmnopqrstuvwxyzàáâãèéêìíîòóôõùúûç'
    for E, D in fatias:
        for letra in letras:
            novas_palavras.append(E + letra + D)
    return novas_palavras

def gerador_palavras(palavra):
    fatias = []
    for i in range(len(palavra)+1):
        fatias.append((palavra[:i],palavra[i:])) # Correção
    palavras_geradas = insere_letras(fatias)
    return palavras_geradas

palavras_geradas = gerador_palavras(palavra_exemplo)
print(palavras_geradas)


['algica', 'blgica', 'clgica', 'dlgica', 'elgica', 'flgica', 'glgica', 'hlgica', 'ilgica', 'jlgica', 'klgica', 'llgica', 'mlgica', 'nlgica', 'olgica', 'plgica', 'qlgica', 'rlgica', 'slgica', 'tlgica', 'ulgica', 'vlgica', 'wlgica', 'xlgica', 'ylgica', 'zlgica', 'àlgica', 'álgica', 'âlgica', 'ãlgica', 'èlgica', 'élgica', 'êlgica', 'ìlgica', 'ílgica', 'îlgica', 'òlgica', 'ólgica', 'ôlgica', 'õlgica', 'ùlgica', 'úlgica', 'ûlgica', 'çlgica', 'lagica', 'lbgica', 'lcgica', 'ldgica', 'legica', 'lfgica', 'lggica', 'lhgica', 'ligica', 'ljgica', 'lkgica', 'llgica', 'lmgica', 'lngica', 'logica', 'lpgica', 'lqgica', 'lrgica', 'lsgica', 'ltgica', 'lugica', 'lvgica', 'lwgica', 'lxgica', 'lygica', 'lzgica', 'làgica', 'lágica', 'lâgica', 'lãgica', 'lègica', 'légica', 'lêgica', 'lìgica', 'lígica', 'lîgica', 'lògica', 'lógica', 'lôgica', 'lõgica', 'lùgica', 'lúgica', 'lûgica', 'lçgica', 'lgaica', 'lgbica', 'lgcica', 'lgdica', 'lgeica', 'lgfica', 'lggica', 'lghica', 'lgiica', 'lgjica', 'lgkica', 'lglica',

Como saída de palavras_geradas, teremos todas as palavras geradas por inserção.

Precisaremos de uma função que indicará 'lógica' como a palavra correta dentre as demais. Na célula seguinte, definiremos corretor() recebendo a palavra.

Quando recebermos a palavra equivocada, chamaremos a função gerador_palavras() com a palavra dentro da variável palavras_geradas. Por fim, o corretor() retornará a palavra_correta.

Em seguida, a grande questão será como definiremos essa palavra_correta.

Temos um arquivo de texto, e queremos verificar se um termo está correto ou não de acordo com sua existência em nossa base de dados. Também deveremos calcular suas probabilidades dentro do conjunto de palavras geradas, pegando a correta como a de máxima probabilidade.

Para entendermos melhor, diremos que a palavra_correta é igual a função max() do Python, o qual retornará o elemento de máxima probabilidade.

Passaremos palavras_geradas para o método, ou seja, as strings do conjunto resultantes da impressão desta variável.

Porém, queremos a probabilidade de cada uma dessas palavras ser a correta. Por exemplo, os resultados 'álgica' ou 'glgica' não possuem chances de estarem certos.

Portanto, a palavra 'lógica' provavelmente terá a maior probabilidade, ainda que seja possível que outro termo existente no vocabulário português tenha se formado com as operações.

Para fazermos este cálculo dentro de max(), chamaremos a função probabilidade através do parâmetro key, a qual será construída mais adiante.

In [29]:
def corretor(palavra):
    palavras_geradas = gerador_palavras(palavra)
    palavra_correta = max(palavras_geradas, key=probabilidade)
    return palavra_correta

Desta maneira, retornaremos a palavra de maior probabilidade.

Neste passo, calcularemos a probabilidade dos resultados gerados.

Pra construirmos a função probabilidade() em uma nova célula, primeiro precisaremos fazer este cálculo com uma palavra; a frequência do termo “lógica” será a quantidade de vezes que este parece no texto em relação ao total de palavras. Para calcular, o nltk possui a função .FreqDist().

Para isso, criaremos uma variável chamada frequência sendo igual ao método que calcula a Distribuição de Frequência dos vocábulos, e passaremos a lista_normalizada como parâmetro.

Na linha seguinte, plotaremos os dez primeiros itens escrevendo frequencia com .most_common() recebendo 10.

In [30]:
frequencia = nltk.FreqDist(lista_normalizada)
frequencia.most_common(10)

[('de', 15502),
 ('o', 14056),
 ('que', 12230),
 ('a', 11099),
 ('e', 10501),
 ('para', 7710),
 ('um', 6367),
 ('é', 5899),
 ('uma', 5220),
 ('do', 5124)]

Executando o código, criaremos uma lista com as dez palavras mais frequentes no corpus. Desta forma, acessaremos a frequência de cada uma utilizando a anotação de dicionário.

Na célula seguinte, escreveremos frequencia[] e passaremos a palavra que queremos dentro dos colchetes, para então descobrirmos a frequência. Neste caso, passaremos o termo “lógica”.

In [31]:
frequencia["lógica"]

96

Com isso, receberemos o valor de 87 vezes em que este vocábulo aparece no corpus.

Em seguida, construiremos a função de probabilidade() com def na próxima célula, e passaremos a palavra_gerada. Apenas relembrando, a função max() calculou a probabilidade chamando probabilidade e passando a palavra gerada pelo gerador_palavras().

O return desta função será a frequencia[] da palavra_gerada dividido pelo total_palavras.

Em uma nova célula anterior a esta, criaremos a variável total_palavras sendo igual ao len() da lista_normalizada, da mesma forma como fizemos anteriormente para descobrirmos o valor de 17654 tipos diferentes de palavras.

In [32]:
total_palavras = len(lista_normalizada)

In [33]:
def probabilidade(palavra_gerada):
    return frequencia[palavra_gerada]/total_palavras

Estamos utilizando uma variável global porque a probabilidade() será chamada constantemente para retornar apenas os valores.

Para melhorarmos a visualização do código, recortaremos a linha de total_palavras e colaremos na mesma célula onde a frequencia foi definida anteriormente.

In [34]:
frequencia = nltk.FreqDist(lista_normalizada)
total_palavras = len(lista_normalizada)
frequencia.most_common(10)

[('de', 15502),
 ('o', 14056),
 ('que', 12230),
 ('a', 11099),
 ('e', 10501),
 ('para', 7710),
 ('um', 6367),
 ('é', 5899),
 ('uma', 5220),
 ('do', 5124)]

Também apagaremos a célula onde fizemos um teste com o termo "lógica" em frequencia[].

Na função probabilidade(), calcularemos a probabilidade da palavra “lógica” no corpus, pois é a correta no caso do equívoco em "lgica".

In [35]:
def probabilidade(palavra_gerada):
    return frequencia[palavra_gerada]/total_palavras

probabilidade("lógica")

0.00023819507680550628

Como retorno, teremos o número 0.00022 aproximadamente. Este valor pequeno se dá por conta do grande volume total de palavras, onde "lógica" aparece 87 vezes.

Na célula seguinte, calcularemos a probabilidade de uma outra palavra, como “lagica” por exemplo, a qual foi gerada como candidata a correção pelo algoritmo de inserção.

In [36]:
probabilidade("lagica")

0.0

Sua probabilidade será 0, afinal não há ocorrências deste termo no corpus. Ainda, a frequência da palavra “logica” sem acento é um número muito menor do que em relação à “lógica” com acento.

Isso acontece porque a autora ou autor do texto eventualmente pode esquecer de adicionar a acentuação correta em uma palavra. Porém isso também é um erro de digitação, e aparecem com bem menos frequência do que a palavra correta.

Portanto, nosso algoritmo será capaz de corrigir termos como "lgica" para “lógica”. Para testarmos, chamaremos a função corretor() com a palavra_exemplo.

In [37]:
corretor(palavra_exemplo)

'lógica'

A saída do corretor será justamente a palavra 'lógica' mais frequente, a qual é a correção para quando digitamos erroneamente “lgica”. Portanto, conseguimos completar o pipeline do corretor ortográfico para este primeiro exemplo.

Porém, só estamos corrigindo o tipo de erro onde uma letra não foi digitada, o termo “lgica”, e é interessante tratarmos outros casos.

Se por acaso digitarmos um caractere a mais, como "lógicaa" ou "lógicca" por exemplo, também deveremos ter outras operações que realizam a correção, além da inserção que já vimos.

# Preparando dados de teste

Nosso corretor já é capaz de corrigir um único tipo de erro, mas precisaremos ampliar os operações de correção para abranger mais casos equivocados.

Para comprovarmos se realmente melhoramos a eficiência do algoritmo, compararemos os resultados do corretor inicial com a versão final.

Porém, ainda precisamos avaliar o nosso modelo, pois só realizamos testes com a palavra “lógica” até agora.

Então definiremos uma função chamada avaliador() que receberá alguns dados de teste, mas ainda não sabemos quais são e os descobriremos adiante. Por enquanto, os chamaremos de testes.

Com isso, receberemos uma porcentagem da taxa de acerto do meu corretor, e a printaremos com a frase "taxa de acerto".

Agora, precisaremos descobrir quais são esses dados recebidos, e para isso veremos a nossa base de dados palavras.txt preparada para teste.

Teremos uma lista com diversas palavras em pares, onde a primeira é a correta e a segunda está errada.

O avaliador() precisará receber a palavra incorreta, passá-la pelo corretor e verificar se este acertou o resultado.

No Google Colab, importaremos os dados e os prepararemos para a função. Antes desta, definiremos cria_dados_teste() recebendo justamente a base de dados palavras.txt que será chamada de nome_arquivo.

Como retorno, teremos uma lista_palavras_teste, a qual será criada sendo igual a uma lista vazia antes de return.

In [38]:
def cria_dados_teste(nome_arquivo):
    lista_palavras_teste = []

    return lista_palavras_teste

def avaliador(testes):
    print("taxa de acerto")

Queremos ler cada uma das linhas do arquivo .txt, separando o lado correto do incorreto. Logo, precisaremos criar duas variáveis: a correta e a errada. Chamaremos essa leitura como linha mesmo.

Para separarmos as palavras por espaço, poderemos utilizar o método split(). Esta linha vem justamente da leitura do arquivo apelidada como f no início do curso.

Manteremos esta mesma nomenclatura para sermos consistentes. Então criaremos a variável f sendo igual a open() recebendo o nome_arquivo. Para que fique aberto para leitura, adicionaremos "r" como segundo parâmetro.

Em seguida, iremos iterar sobre o arquivo e leremos cada uma de suas linhas, fazendo um for em linha dentro de f. Quando realizarmos a iteração, transformaremos cada linha em uma string.

Então, para cada uma delas no meu arquivo f, faremos a divisão entre correta e errada com .split() como já escrevemos.

Realizaremos o .append() na lista_palavras_teste também, passando uma tupla onde o primeiro lado será a palavra correta e o segundo, a errada.

Por fim, fecharemos o arquivo com f.close().

In [39]:
def cria_dados_teste(nome_arquivo):
    lista_palavras_teste = []
    f = open(nome_arquivo, "r")
    for linha in f:
        correta, errada = linha.split()
        lista_palavras_teste.append((correta, errada))
    f.close()
    return lista_palavras_teste

def avaliador(testes):
    print("taxa de acerto")

Com isso, criamos a função que pegará os nossos dados e os preparará para avaliação.

Falta-nos enviar a base de dados para as máquinas do Google, pois ainda não está disponível; expandiremos a aba lateral do Colab para fazermos o upload de palavras.txt. Ao fim do carregamento, conseguiremos utilizar o arquivo de fato.

Em seguida, criaremos a variável lista_teste no lugar da definição do avaliador() que será apagado por enquanto. Essa lista receberá o retorno da função cria_dados_teste() com o "palavras.txt".

Com isso, teremos as tuplas e poderemos avaliar nosso corretor com os dados da forma que precisamos.

In [40]:
def cria_dados_teste(nome_arquivo):
    lista_palavras_teste = []
    f = open(nome_arquivo, "r")
    for linha in f:
        correta, errada = linha.split()
        lista_palavras_teste.append((correta, errada))
    f.close()
    return lista_palavras_teste

lista_teste = cria_dados_teste("palavras.txt")
lista_teste

[('podemos', 'pyodemos'),
 ('esse', 'esje'),
 ('já', 'jrá'),
 ('nosso', 'nossov'),
 ('são', 'sãêo'),
 ('dos', 'dosa'),
 ('muito', 'muifo'),
 ('imagem', 'iômagem'),
 ('sua', 'ósua'),
 ('também', 'tambéùm'),
 ('ele', 'eme'),
 ('fazer', 'èazer'),
 ('temos', 'temfs'),
 ('essa', 'eàssa'),
 ('quando', 'quaôdo'),
 ('vamos', 'vamvos'),
 ('sobre', 'hsobre'),
 ('java', 'sjava'),
 ('das', 'daõs'),
 ('agora', 'agorah'),
 ('está', 'eòtá'),
 ('cada', 'céda'),
 ('mesmo', 'zmesmo'),
 ('nos', 'noâ'),
 ('forma', 'fobma'),
 ('seja', 'sejéa'),
 ('então', 'enêão'),
 ('criar', 'èriar'),
 ('código', 'cóeigo'),
 ('caso', 'casío'),
 ('exemplo', 'áexemplo'),
 ('tem', 'tĩem'),
 ('usuário', 'usuárôio'),
 ('dados', 'dfados'),
 ('python', 'pgthon'),
 ('nossa', 'nossah'),
 ('além', 'alémè'),
 ('assim', 'asõim'),
 ('ter', 'teb'),
 ('até', 'atĩ'),
 ('bem', 'âem'),
 ('design', 'desigen'),
 ('trabalho', 'trabalàho'),
 ('foi', 'foo'),
 ('apenas', 'apenaũ'),
 ('empresa', 'empresà'),
 ('valor', 'valíor'),
 ('será', 'serr')

Ao executarmos a célula, teremos a nossa base de dados em uma estrutura que facilita a criação do avaliador().

Neste passo, desenvolveremos a função avaliador().

A criaremos com def e passaremos os testes gerados pela função cria_dados_teste, para então printarmos a "taxa de acerto" com a porcentagem de quantas palavras o corretor está acertando de fato.

In [41]:
def avaliador (testes):
    print("taxa de acerto")

O primeiro passo será calcular justamente esta taxa com a variável taxa_acerto, a qual será igual a acertou sobre numero_palavras, correspondente ao número total.

Em seguida, precisaremos descobrir o número de palavras e a quantidade de acertos. O primeiro elemento será igual ao tamanho da variável teste que estamos recebendo no avaliador(), pois o len() de testes será este valor total.

Já em acertou, precisaremos criar um contador começando do zero, então será igual a 0. Este elemento será implementado toda vez que o corretor acertar.

Precisaremos iterar sobre cada uma das linhas da lista_testes, passar a palavra equivocada para a função do corretor e verificar se a que foi gerada é igual à correta. Em caso positivo, acrescentaremos 1 à variável acertou.

Para isso, desenvolveremos uma lógica; para cada for das linhas, retornaremos a palavra correta e a errada de testes.

Passaremos a equivocada para o corretor() já desenvolvido, o qual retornará a palavra_corrigida. Caso esta seja igual à correta em if, implementaremos o 1 ao acertou com +=.

Depois disso, já teremos a taxa de acerto em acertou sobre o número_palavras.

No print() alteraremos o conteúdo "taxa de acerto" para a variável taxa_acerto dentro de aspas e chaves, fazendo a interpolação de string com f. Melhoraremos a visualização do código incluindo % e o número_palavras também.

Para calcularmos a porcentagem, multiplicaremos acertou por 100 em taxa_acerto. Para arredondarmos o resultado, usaremos a função round() recebendo esta operação da taxa de acerto e o número de casas decimais que queremos, sendo 2 neste caso.

Com isso, poderemos testar nosso corretor chamando a função avaliador() recebendo a lista_teste.

In [42]:
def avaliador (testes):
    numero_palavras = len(testes)
    acertou = 0
    for correta, errada in testes:
        palavra_corrigida = corretor(errada)
        if palavra_corrigida == correta:
            acertou += 1
    taxa_acerto = acertou/numero_palavras
    print(f'{taxa_acerto}')

avaliador(lista_teste)

0.010752688172043012


Para deixar mais legível 

In [43]:
def avaliador (testes):
    numero_palavras = len(testes)
    acertou = 0
    for correta, errada in testes:
        palavra_corrigida = corretor(errada)
        if palavra_corrigida == correta:
            acertou += 1
            
# round arredondando e dizendo a quantidade da casa decimal 2

    taxa_acerto = round(acertou*100/numero_palavras, 2)
    print(f"{taxa_acerto}% de {numero_palavras} palavras")

avaliador(lista_teste)

1.08% de 186 palavras


Ao rodarmos a célula, teremos o resultado de apenas 1.08% de taxa de acerto num total de 186 palavras.

Este é um número inadmissível, pois é muito baixo para um corretor ortográfico.

Como já conseguimos comparar os nossos corretores, sabemos o quanto estamos acertando para podermos melhorá-lo.

A operação que realizamos para corrigir o primeiro problema da falta de um caractere em "lgica" insere uma letra em todas as posições possíveis da palavra, gerando vários resultados.

Neste passo, corrigiremos um outro tipo de erro bastante comum também além de deletar ou esquecer de digitar, que consiste em escrever uma letra a mais por engano.

Por exemplo, nosso algoritmo receberá o termo equivocado “lóigica”, e notaremos que a tecla “O” fica ao lado de “I” no teclado padrão, o que facilmente poderia levar a este erro. Logo, precisaremos que nos retorne a palavra correta “lógica”.

Para construirmos este novo algoritmo, continuaremos com o sistema de dividir a palavra em dois lados como fizemos anteriormente.

Na primeira linha de divisão na tabela, teremos a primeira parte da palavra null e a segunda "lóigica" para realizarmos a operação, a qual sempre pegará o lado esquerdo e somará com algo.

Porém, há uma letra a mais que precisa ser eliminada. Portanto, deletaremos o primeiro caractere da segunda parte, obtendo null + "óigica". Concatenaremos pela operação de soma para gerar o resultado. Neste caso, "óigica" mesmo.

No segundo caso, o lado esquerdo é o “L” e o direito é o “óigica”, e realizaremos a mesma operação eliminando a primeira letra da segunda parte da palavra, resultando em “ligica” que também está equivocada.

Como terceiro exemplo, teremos a primeira parte com “Ló” e a segunda “igica”. Eliminando a letra "i" escrita a mais por engano neste lado direito, teremos finalmente o resultado correto "lógica".

Com isso, criamos um novo algoritmo que chegou à resposta certa. Continuaremos a tabela até o último caso para vermos os demais retornos.

LÓIGICA

<table><thead><tr><th>ESQUERDO</th><th>DIREITO</th><th>OPERAÇÃO</th><th>RESULTADO</th></tr></thead><tbody><tr><td>null</td><td><del>l</del>óigica</td><td>null + óigica</td><td>óigica</td></tr><tr><td>l</td><td><del>ó</del>igica</td><td>l + igica</td><td>ligica</td></tr><tr><td>ló</td><td><del>i</del>gica</td><td>ló + gica</td><td><strong>lógica</strong></td></tr><tr><td>lói</td><td><del>g</del>ica</td><td>lói + ica</td><td>lóiica</td></tr><tr><td>lóig</td><td><del>i</del>ca</td><td>lóig + ca</td><td>lóigca</td></tr><tr><td>lóigi</td><td><del>c</del>a</td><td>lóigi + a</td><td>lóigia</td></tr><tr><td>lóigic</td><td><del>a</del></td><td>lóigic +</td><td>lóigic</td></tr><tr><td>lóigica</td><td>null</td><td>lóigica + null</td><td>lóigica</td></tr></tbody></table>

Com isso, conseguimos elaborar um algoritmo que resolve nosso problema.

Precisaremos generalizar e pensar em qual operação estamos realizando; eliminamos a primeira letra do lado direito e somamos as duas partes. A soma em Python é a concatenação, então estamos concatenando os dois pedaços de string.

ESQUERDO + DIREITO[1:]

Como a segunda parte da palavra inicia com o caractere de index 1, sempre eliminaremos o de index 0.

A seguir, implementaremos este novo método para vermos como nosso corretor se comportará.

# Deletando caracteres

Para gerarmos o novo algoritmo com base na tabela do passo anterior, criaremos a função deletando_caracteres() com def em uma nova célula, pois pode ser o caractere equivocado digitado a mais não seja sempre uma letra, e sim uma pontuação por exemplo.

Esta receberá as fatias com os lados esquerdo e direito, as percorrerá e realizará a operação que queremos, da mesma forma feita com a anterior.

Copiaremos e colaremos o corpo de insere_letras() para fazermos algumas adaptações a esta nova situação.

Já não precisaremos mais das letras, pois não inseriremos nenhuma neste caso de erro, então poderemos deletar esta linha.

A variável novas_palavras será o retorno com as novas palavras geradas.

Com o for, pegaremos os dois lados e não precisaremos mais do for interno, pois já teremos a operação.

A palavra que será appendada na lista com append() será a operação com a soma das duas partes E e D, mas desta vez iniciando no index [1:] como já vimos.

In [44]:
def deletando_caracteres(fatias):
    novas_palavras = []
    for E, D in fatias:
        novas_palavras.append(E + D[1:])
    return novas_palavras

Por exemplo, digitaremos "palavras"[] na célula seguinte, pegando do index 1 até o último, o que retornará o resultado “alavra”.

Logo, isso eliminará a primeira letra como queremos.

Para testarmos a função, chamaremos gerador_palavras() para gerarmos diversas outras palavras com os algoritmos.

Relembrando-a, pegamos a palavra, criamos as fatias divididas entre os dois lados e geramos a variável palavras_geradas.

Porém, esta só chama a função insere_letras() por enquanto, e precisaremos adaptar a gerador_palavras() com o novo algoritmo que elimina caracteres.

A seguir, faremos a adaptação para podermos testar a função e verificar se conseguiremos gerar a palavra correta.

Como gerador_palavras() ainda não está implementado para gerar as palavras através da função deletando_caracteres, é capaz de corrigir apenas o primeiro caso de uma letra a menos.

Se colocarmos a palavra “lóigica” do novo caso de erro com um caractere a mais no gerador, a função retornará as palavras geradas, sendo a primeira “alóigica”.

Logo, precisaremos adaptar o gerador_palavras(); copiaremos todo seu bloco e colaremos em uma nova célula. Esta função fatia a string, gera uma lista de novas palavras e a armazena em palavras_geradas que será retornada. Então precisaremos gerar uma nova com deletando_caracteres().

Como é uma lista também, concatenaremos com a anterior. Pegaremos a variável palavras_geradas do insere_letras() e somaremos com a lista gerada pelo deletando_caracteres() recebendo as fatias.

Feito isso, o gerador_palavras() estará adaptado para gerar as palavras também através do outro algoritmo. Em seguida, verificaremos se gerará a palavra correta com o exemplo "lóigica".

a variável palavra_exemplo recebida como parâmetro de gerador_palavras() será criada na linha anterior, sendo igual a "lóigica".

Por fim, imprimiremos as palavras geradas.

In [45]:
def gerador_palavras(palavra):
    fatias = []
    for i in range(len(palavra)+1):
        fatias.append((palavra[:i],palavra[i:]))
    palavras_geradas = insere_letras(fatias)
    palavras_geradas += deletando_caracteres(fatias)
    return palavras_geradas
palavra_exemplo = "lóigica"
palavras_geradas = gerador_palavras(palavra_exemplo)
print(palavras_geradas)

['alóigica', 'blóigica', 'clóigica', 'dlóigica', 'elóigica', 'flóigica', 'glóigica', 'hlóigica', 'ilóigica', 'jlóigica', 'klóigica', 'llóigica', 'mlóigica', 'nlóigica', 'olóigica', 'plóigica', 'qlóigica', 'rlóigica', 'slóigica', 'tlóigica', 'ulóigica', 'vlóigica', 'wlóigica', 'xlóigica', 'ylóigica', 'zlóigica', 'àlóigica', 'álóigica', 'âlóigica', 'ãlóigica', 'èlóigica', 'élóigica', 'êlóigica', 'ìlóigica', 'ílóigica', 'îlóigica', 'òlóigica', 'ólóigica', 'ôlóigica', 'õlóigica', 'ùlóigica', 'úlóigica', 'ûlóigica', 'çlóigica', 'laóigica', 'lbóigica', 'lcóigica', 'ldóigica', 'leóigica', 'lfóigica', 'lgóigica', 'lhóigica', 'lióigica', 'ljóigica', 'lkóigica', 'llóigica', 'lmóigica', 'lnóigica', 'loóigica', 'lpóigica', 'lqóigica', 'lróigica', 'lsóigica', 'ltóigica', 'luóigica', 'lvóigica', 'lwóigica', 'lxóigica', 'lyóigica', 'lzóigica', 'làóigica', 'láóigica', 'lâóigica', 'lãóigica', 'lèóigica', 'léóigica', 'lêóigica', 'lìóigica', 'líóigica', 'lîóigica', 'lòóigica', 'lóóigica', 'lôóigica', 'lõ

Como saída, teremos os resultados de insere_letras() seguido dos de deletando_caracteres(), indicando que a função está funcionando corretamente. Dentro das listas retornadas, encontraremos a palavra correta “lógica”.

Na célula seguinte, chamaremos a função avaliador() recebendo a lista_teste.

In [46]:
avaliador(lista_teste)

41.4% de 186 palavras


No retorno deste comando, teremos uma taxa de acerto de 41.4% de 186 palavras. Ou seja, fizemos um ótimo upgrade em comparação com com o resultado anterior.

Este já é um retorno bem mais interessante, mas ainda podemos melhorar ainda mais como veremos a seguir.

# Trocando as letras

Nosso corretor já é capaz de resolver dois tipos de erros de digitação.

Resumidamente, as operações consistem em dividir a string equivocada em duas partes e adicionar ou eliminar um caractere para obter diferentes resultados até chegar na palavra correta, através das funções insere_letras() e deletando_caracteres() respectivamente.

Porém, ainda existem outros tipos de erros, como quando trocamos uma letra por outra. Por exemplo, na hora de digitar o “ó” no teclado, sem querer apertamos a tecla “í” que está bem ao lado do caractere correto, gerando a palavra errada "lígica".

Logo, precisaremos desenvolver o corretor para realizarmos a operação de troca de letras. Para isso, elaboraremos outra tabela para entendermos melhor como fazê-la.

Começaremos dividindo a palavra "lígica" em duas partes como já sabemos. Na primeira linha, teremos "null" somado com "lígica" mesmo, e neste caso, substituiremos a letra errada e concatenaremos com a correta em algum nos resultados finais da operação.

Esta pegará o primeiro lado, somará com a letra "ó" que é a correta neste caso de erro de digitação, e adicionará o segundo lado.

Como precisaremos trocar as letras, pegaremos um dos caracteres do lado direito e trocaremos pela letra “ó”, eliminando a primeira letra da segunda parte da palavra e concatenando com o restante da string.

Com isso, o primeiro resultado será o termo “óigica” que ainda está equivocado. No segundo exemplo da linha seguinte da tabela, a operação será "L + ó + gica", formando enfim a palavra correta "lógica".

Faremos até obtermos todos os resultados possíveis na tabela.

LÍGICA

<table><thead><tr><th>ESQUERDO</th><th>DIREITO</th><th>OPERAÇÃO</th><th>RESULTADO</th></tr></thead><tbody><tr><td>null</td><td><del>l</del>ígica</td><td>null + ó + ígica</td><td>óigica</td></tr><tr><td>l</td><td><del>í</del>gica</td><td>l + ó + gica</td><td><strong>lógica</strong></td></tr><tr><td>lí</td><td><del>g</del>ica</td><td>lí + ó + ica</td><td>lióica</td></tr><tr><td>líg</td><td><del>i</del>ca</td><td>líg + ó + ca</td><td>ligóca</td></tr><tr><td>lígi</td><td><del>c</del>a</td><td>lígi + ó + a</td><td>ligióa</td></tr><tr><td>lígic</td><td><del>a</del></td><td>lígic + ó</td><td>ligicó</td></tr><tr><td>lígica</td><td>null</td><td>lígica + ó + null</td><td>ligicaó</td></tr></tbody></table>

Com isso, teremos o primeiro lado, mais a letra, mais o segundo lado, menos a primeira letra desta mesma segunda parte.

ESQUERDO + LETRA + DIREITO[1:]

Notaremos que essa operação é praticamente a soma das anteriores, insere_letras() e deletando_caracteres(). Então, a nova função que trocará as letras ficará bastante parecida com estas duas operações que nós já realizamos, o que facilitará bastante o desenvolvimento dos algoritmos.

A seguir, faremos esta implementação.

Para resolvermos o problema de digitação anterior, criaremos a função troca_letra(), que substituirá uma letra equivocada pela correta.

No caso do termo "lígica" escrito erroneamente, nosso algoritmo precisará pegar a letra “í” e substituir por “ó”, corrigindo para "lógica".

A nova função receberá as fatias e, como já comentamos, será a junção das funções insere_letras() e deletando_caracteres() criadas anteriormente.

Então poderemos utilizar as funções que já implementamos para escrevermos a troca_letra(); copiaremos o corpo de insere_letras() e colaremos dentro da nova função para fazermos as adaptações cabíveis.

Primeiro, precisaremos deletar a primeira letra de index 0 da segunda parte da palavra D, então pegaremos do index 1 em diante como fizemos em deletando_caracteres().

In [47]:
def troca_letra(fatias):
    novas_palavras = []
    letras = 'abcdefghijklmnopqrstuvwxyzáàãâéèêíìóòõôúùûç'
    for E, D in fatias:
        for letra in letras:
            novas_palavras.append(E + letra + D[1:])
    return novas_palavras

Com isso, teremos a operação de troca_letra() implementada.

Antes de prosseguirmos, melhoraremos nosso código, pois estas operações, o gerador de palavras e o avaliador estão muito distantes entre si. Então os colocaremos em uma única célula para facilitar as modificações.

Em seguida, precisaremos chamar a nova função troca_letra() recebendo as fatias dentro de gerador_palavras().

Ao final, teremos a palavra_exemplo sendo igual a "lígica" e imprimiremos as palavras_geradas.

In [48]:
def insere_letras(fatias):
    novas_palavras = []
    letras = 'abcdefghijklmnopqrstuvwxyzàáâãèéêìíîòóôõùúûç'
    for E, D in fatias:
        for letra in letras:
            novas_palavras.append(E + letra + D)
    return novas_palavras

def deletando_caracteres(fatias):
    novas_palavras = []
    for E, D in fatias:
        novas_palavras.append(E + D[1:])
    return novas_palavras

def troca_letra(fatias):
    novas_palavras = []
    letras = 'abcdefghijklmnopqrstuvwxyzáàãâéèêíìóòõôúùûç'
    for E, D in fatias:
        for letra in letras:
            novas_palavras.append(E + letra + D[1:])
    return novas_palavras

def gerador_palavras(palavra):
    fatias = []
    for i in range(len(palavra)+1):
        fatias.append((palavra[:i],palavra[i:]))
    palavras_geradas = insere_letras(fatias)
    palavras_geradas += deletando_caracteres(fatias)
    palavras_geradas += troca_letra(fatias)
    return palavras_geradas

palavra_exemplo = "lígica"
palavras_geradas = gerador_palavras(palavra_exemplo)
print(palavras_geradas)

['alígica', 'blígica', 'clígica', 'dlígica', 'elígica', 'flígica', 'glígica', 'hlígica', 'ilígica', 'jlígica', 'klígica', 'llígica', 'mlígica', 'nlígica', 'olígica', 'plígica', 'qlígica', 'rlígica', 'slígica', 'tlígica', 'ulígica', 'vlígica', 'wlígica', 'xlígica', 'ylígica', 'zlígica', 'àlígica', 'álígica', 'âlígica', 'ãlígica', 'èlígica', 'élígica', 'êlígica', 'ìlígica', 'ílígica', 'îlígica', 'òlígica', 'ólígica', 'ôlígica', 'õlígica', 'ùlígica', 'úlígica', 'ûlígica', 'çlígica', 'laígica', 'lbígica', 'lcígica', 'ldígica', 'leígica', 'lfígica', 'lgígica', 'lhígica', 'liígica', 'ljígica', 'lkígica', 'llígica', 'lmígica', 'lnígica', 'loígica', 'lpígica', 'lqígica', 'lrígica', 'lsígica', 'ltígica', 'luígica', 'lvígica', 'lwígica', 'lxígica', 'lyígica', 'lzígica', 'làígica', 'láígica', 'lâígica', 'lãígica', 'lèígica', 'léígica', 'lêígica', 'lìígica', 'líígica', 'lîígica', 'lòígica', 'lóígica', 'lôígica', 'lõígica', 'lùígica', 'lúígica', 'lûígica', 'lçígica', 'líagica', 'líbgica', 'lícgica'

Rodando esta célula, teremos todas as as palavras geradas pelos algoritmos, inclusive pela troca de letras, e dentre elas a correta "lógica".

Portanto, a função troca_letra() está funcionando conforme o esperado.

Mas ainda existe mais um tipo de erro para corrigirmos, e o veremos a seguir para podermos testar o algoritmo final do corretor.

# Invertendo as letras

Neste passo, corrigiremos um erro de digitação onde letras foram invertidas em suas posições, como na palavra escrita errada "lgóica".

Neste caso, a posição da letra “g” foi trocada com a letra “ó”, e como se trata de um equívoco bastante comum, precisaremos deixar o corretor cada vez mais abrangente.

Portanto, implementaremos um algoritmo capaz de realizar esse tipo de correção.

Precisaremos inverter novamente o caractere “ó” com a letra “g”; logo, esta operação será a inversão das letras.

Entenderemos o passo a passo da construção deste algoritmo com uma nova tabela da mesma forma que fizemos com os anteriores.

Dividiremos a string "lgóica" em duas partes como já nos habituamos. Para começarmos, pegaremos o primeiro lado null da primeira linha e realizaremos a concatenação com a segunda letra "g".

Depois, pegaremos a primeira letra do segundo lado "l" e realizaremos a concatenação com o restante da palavra "óica" neste primeiro resultado.

Em resumo, pegaremos o lado esquerdo, somaremos com a segunda letra do lado direito, depois concatenaremos com a primeira letra deste mesmo lado direito e por fim somaremos com o restante da palavra.

O resultado desta primeira operação será a palavra "glóica", que obviamente está errada também.

Seguiremos por todas as linhas da tabela até o último caso para vermos se a palavra correta "lógica" aparece em algum dos resultados.

LGÓICA

<table><thead><tr><th>ESQUERDO</th><th>DIREITO</th><th>OPERAÇÃO</th><th>RESULTADO</th></tr></thead><tbody><tr><td>null</td><td>lgóica</td><td>null + g + l + óica</td><td>glóica</td></tr><tr><td>l</td><td>góica</td><td>l + ó + g + ica</td><td><strong>lógica</strong></td></tr><tr><td>lg</td><td>óica</td><td>lg + i + ó + ca</td><td>lgióca</td></tr><tr><td>lgó</td><td>ica</td><td>lgó + c + i + a</td><td>lgócia</td></tr><tr><td>lgói</td><td>ca</td><td>lgói + a + c + null</td><td>lgóiac</td></tr><tr><td>lgóic</td><td>a</td><td>lgóic + a + a + null</td><td>lgóicaa</td></tr><tr><td>lgóica</td><td>null</td><td>lgóica + null + null + null</td><td>lgóica</td></tr></tbody></table>

Feito isso, já sabemos qual operação este algoritmo será capaz de realizar para a inversão de letras em toda minha palavra.

Esta será composta pelo lado esquerdo, somado com a segunda letra de index 1 do lado direito, visto que sempre começamos com 0. Ainda somaremos a primeira letra de index 0 da segunda parte com restante da palavra, ou seja, da terceira posição de index 2 em diante.

ESQUERDO + DIREITO[1] + DIREITO[0] + DIREITO[2:]

Agora que sabemos como realizar esta operação, poderemos implementá-la adiante.

Implementaremos o algoritmo de inversão de letras neste passo.

Desenvolveremos a nova função na mesma célula onde já estão todas as operações do nosso corretor.

Nesta, notaremos que temos a mesma variável letras em duas das nossas operações; na insere_letras() e em troca_letra() também.

Porém, não precisamos desta nem em deletando_caracteres(), e nem precisaremos na nova função inverte_letra() criada para o erro de digitação de "lgóica".

Com isso, utilizaremos as vantagens de já termos dividido a string em dois lados. Praticamente copiaremos o corpo de deletando_letras() e colaremos no novo algoritmo logo após troca_letra() para alterarmos a operação.

Dentro de append(), inseriremos a operação E + D[1] + D[0] + D[2:] como vimos anteriormente.

In [49]:
def inverte_letra(fatias):
    novas_palavras = []
    for E, D in fatias:
        novas_palavras.append(E + D[1] + D[0] + D[2:])
    return novas_palavras


In [50]:
def insere_letras(fatias):
    novas_palavras = []
    letras = 'abcdefghijklmnopqrstuvwxyzàáâãèéêìíîòóôõùúûç'
    for E, D in fatias:
        for letra in letras:
            novas_palavras.append(E + letra + D)
    return novas_palavras

def deletando_caracteres(fatias):
    novas_palavras = []
    for E, D in fatias:
        novas_palavras.append(E + D[1:])
    return novas_palavras

def troca_letra(fatias):
    novas_palavras = []
    letras = 'abcdefghijklmnopqrstuvwxyzáàãâéèêíìóòõôúùûç'
    for E, D in fatias:
        for letra in letras:
            novas_palavras.append(E + letra + D[1:])
    return novas_palavras

def inverte_letra(fatias):
    novas_palavras = []
    for E, D in fatias:
        novas_palavras.append(E + D[1] + D[0] + D[2:])
    return novas_palavras

def gerador_palavras(palavra):
    fatias = []
    for i in range(len(palavra)+1):
        fatias.append((palavra[:i],palavra[i:]))
    palavras_geradas = insere_letras(fatias)
    palavras_geradas += deletando_caracteres(fatias)
    palavras_geradas += troca_letra(fatias)
    return palavras_geradas

palavra_exemplo = "lígica"
palavras_geradas = gerador_palavras(palavra_exemplo)
print(palavras_geradas)

['alígica', 'blígica', 'clígica', 'dlígica', 'elígica', 'flígica', 'glígica', 'hlígica', 'ilígica', 'jlígica', 'klígica', 'llígica', 'mlígica', 'nlígica', 'olígica', 'plígica', 'qlígica', 'rlígica', 'slígica', 'tlígica', 'ulígica', 'vlígica', 'wlígica', 'xlígica', 'ylígica', 'zlígica', 'àlígica', 'álígica', 'âlígica', 'ãlígica', 'èlígica', 'élígica', 'êlígica', 'ìlígica', 'ílígica', 'îlígica', 'òlígica', 'ólígica', 'ôlígica', 'õlígica', 'ùlígica', 'úlígica', 'ûlígica', 'çlígica', 'laígica', 'lbígica', 'lcígica', 'ldígica', 'leígica', 'lfígica', 'lgígica', 'lhígica', 'liígica', 'ljígica', 'lkígica', 'llígica', 'lmígica', 'lnígica', 'loígica', 'lpígica', 'lqígica', 'lrígica', 'lsígica', 'ltígica', 'luígica', 'lvígica', 'lwígica', 'lxígica', 'lyígica', 'lzígica', 'làígica', 'láígica', 'lâígica', 'lãígica', 'lèígica', 'léígica', 'lêígica', 'lìígica', 'líígica', 'lîígica', 'lòígica', 'lóígica', 'lôígica', 'lõígica', 'lùígica', 'lúígica', 'lûígica', 'lçígica', 'líagica', 'líbgica', 'lícgica'

Porém, estamos invertendo o elemento de index 0 com o elemento de index 1. Logo, precisaremos de no mínimo duas letras na segunda fatia, pois do contrário não teremos inversão alguma no caso de termos menos que isso ou null.

Portanto, precisaremos colocar uma condicional if de len() dentro de for do inverte_letra().

Se o tamanho do lado direito D for maior que 1:, realizaremos a operação. Caso seja menor, não fará sentido realizá-la como já discutimos.

Adicionaremos a nova função dentro de gerador_palavras() e alteraremos a palavra_exemplo para "lgóica". Desta forma, poderemos verificar a saída deste algoritmo.

In [51]:
def insere_letras(fatias):
    novas_palavras = []
    letras = 'abcdefghijklmnopqrstuvwxyzàáâãèéêìíîòóôõùúûç'
    for E, D in fatias:
        for letra in letras:
            novas_palavras.append(E + letra + D)
    return novas_palavras

def deletando_caracteres(fatias):
    novas_palavras = []
    for E, D in fatias:
        novas_palavras.append(E + D[1:])
    return novas_palavras

def troca_letra(fatias):
    novas_palavras = []
    letras = 'abcdefghijklmnopqrstuvwxyzáàãâéèêíìóòõôúùûç'
    for E, D in fatias:
        for letra in letras:
            novas_palavras.append(E + letra + D[1:])
    return novas_palavras

def inverte_letra(fatias):
    novas_palavras = []
    for E, D in fatias:
        if len(D) > 1:
            novas_palavras.append(E + D[1] + D[0] + D[2:])
    return novas_palavras

def gerador_palavras(palavra):
    fatias = []
    for i in range(len(palavra)+1):
        fatias.append((palavra[:i],palavra[i:]))
    palavras_geradas = insere_letras(fatias)
    palavras_geradas += deletando_caracteres(fatias)
    palavras_geradas += troca_letra(fatias)
    palavras_geradas += inverte_letra(fatias)
    return palavras_geradas

palavra_exemplo = "lgóica"
palavras_geradas = gerador_palavras(palavra_exemplo)
print(palavras_geradas)

['algóica', 'blgóica', 'clgóica', 'dlgóica', 'elgóica', 'flgóica', 'glgóica', 'hlgóica', 'ilgóica', 'jlgóica', 'klgóica', 'llgóica', 'mlgóica', 'nlgóica', 'olgóica', 'plgóica', 'qlgóica', 'rlgóica', 'slgóica', 'tlgóica', 'ulgóica', 'vlgóica', 'wlgóica', 'xlgóica', 'ylgóica', 'zlgóica', 'àlgóica', 'álgóica', 'âlgóica', 'ãlgóica', 'èlgóica', 'élgóica', 'êlgóica', 'ìlgóica', 'ílgóica', 'îlgóica', 'òlgóica', 'ólgóica', 'ôlgóica', 'õlgóica', 'ùlgóica', 'úlgóica', 'ûlgóica', 'çlgóica', 'lagóica', 'lbgóica', 'lcgóica', 'ldgóica', 'legóica', 'lfgóica', 'lggóica', 'lhgóica', 'ligóica', 'ljgóica', 'lkgóica', 'llgóica', 'lmgóica', 'lngóica', 'logóica', 'lpgóica', 'lqgóica', 'lrgóica', 'lsgóica', 'ltgóica', 'lugóica', 'lvgóica', 'lwgóica', 'lxgóica', 'lygóica', 'lzgóica', 'làgóica', 'lágóica', 'lâgóica', 'lãgóica', 'lègóica', 'légóica', 'lêgóica', 'lìgóica', 'lígóica', 'lîgóica', 'lògóica', 'lógóica', 'lôgóica', 'lõgóica', 'lùgóica', 'lúgóica', 'lûgóica', 'lçgóica', 'lgaóica', 'lgbóica', 'lgcóica'

Executando a célula, veremos os resultados desta última operação ao final da lista impressa. Dentre estes, encontraremos a resposta correta 'lógica'.

Portanto, a função inverte_letra() está implementada corretamente.

Faremos mais um teste com outra palavra equivocadamente escrita com letras invertidas, como "lógiac" por exemplo.

Como resultado desta execução, teremos a impressão da lista com a última palavra sendo a correta 'lógica'. Ou seja, também corrigiu esta inversão de letras com sucesso.

Estando os algoritmos funcionando como o esperado, avaliaremos a eficiência do nosso corretor.

Para avaliarmos o nosso modelo, chamaremos a função avaliador() e passaremos uma lista_teste em uma nova célula seguinte.

In [52]:
avaliador(lista_teste)

76.34% de 186 palavras


O resultado será de 76.34% de 186 palavras. Portanto, melhoramos consideravelmente o nosso corretor, pois começamos com aproximadamente 1% de taxa de acerto apenas, o que era inaceitável.

Porém, estamos realizando os testes a partir de uma lista de palavras que pode conter alguns termos ainda desconhecidos pelo nosso vocabulário.

Então os quase 24% restantes de erro estão associados tanto ao erro do próprio corretor quanto ao fato de desconhecermos a palavra, pois se o algoritmo ainda não a conhece, a correção não será feita.

Portanto, precisaremos saber de alguma maneira qual é a porcentagem de vocábulos conhecidos na lista de teste. Veremos isso a seguir.

#  Palavras desconhecidas ao vocabulário

Nosso corretor já é capaz de acertar aproximadamente 77% das palavras.

A taxa de erros está associada a dois fatores; restrição do algoritmo que não gera a palavra correta e dados de treinamento.

Às vezes um determinado termo ainda é desconhecido, e portanto não podemos realizar a correção de algo que não conhecemos.

Em um exemplo dado anteriormente, se conhecemos cerca de apenas 100 palavras de um idioma estrangeiro, somente poderemos corrigir esta mesma quantidade de palavras conhecidas.

Logo, o modelo do corretor é restrito por este volume de vocabulário, calculado em 18.464 vocábulos em nosso caso.

Precisaremos calcular o quanto que essa restrição está impactando no resultado, pois temos uma taxa de erro e deveremos saber o quanto desta porcentagem está associada às palavras desconhecidas.

In [53]:
def avaliador(testes):
    numero_palavras = len(testes)
    acertou = 0
    for correta, errada in testes:
        palavra_corrigida = corretor(errada)
        if palavra_corrigida == correta:
            acertou += 1
    taxa_acerto = round(acertou*100/numero_palavras, 2)
    print(f"{taxa_acerto}% de {numero_palavras} palavras")

avaliador(lista_teste)

76.34% de 186 palavras


Como estamos calculando a nossa taxa de acerto, poderemos calcular também a taxa de palavras desconhecidas por meio de uma nova variável chamada taxa_desconhecida. Esta será igual a algo muito parecido com o cálculo da taxa de acerto.

Porém, não queremos mais calcular a acertou, e sim o que é desconhecido; então substituiremos pela nova variável desconhecida. Na função, setaremos seu valor com um contador de palavras desconhecidas, da mesma forma que fizemos com acertou igual a 0.

Incrementaremos este contador a cada vez que descobrirmos que uma das palavras corretas em correta é desconhecida ao nosso vocabulário.

Esta implementação entrará no if se acertou o termo. Já o else será justamente quando errar a previsão, ou seja, incrementaremos a variável desconhecida quando a palavra correta for desconhecida do meu vocabulário. Se errou, poderá ser por algum dos dois motivos já citados.

Após +=, será incrementada quando a palavra correta não estiver no vocabulario com not in entre parênteses, e precisaremos recebê-lo como parâmetro de avaliador() também.

Com isso, recebemos o vocabulario como parâmetro do avaliador, incrementamos a variável desconhecida quando a palavra correta não estiver no meu vocabulário e calculamos a taxa de taxa_desconhecida

Falta-nos imprimi-la também na frase e setar a variável vocabulario sendo igual a set() recebendo a lista_normalizada.

Ao final do código, passaremos vocabulario como segundo parâmetro da chamada de avaliador().

In [54]:
def avaliador(testes, vocabulario):
    numero_palavras = len(testes)
    acertou = 0
    desconhecida = 0
    for correta, errada in testes:
        palavra_corrigida = corretor(errada)
        if palavra_corrigida == correta:
            acertou += 1
        else:
            desconhecida += (correta not in vocabulario)
    taxa_acerto = round(acertou*100/numero_palavras, 2)
    taxa_desconhecida = round(desconhecida*100/numero_palavras, 2)
    print(f"{taxa_acerto}% de {numero_palavras} palavras, desconhecidas é {taxa_desconhecida}%")


vocabulario = set(lista_normalizada)
avaliador(lista_teste, vocabulario)

76.34% de 186 palavras, desconhecidas é 6.99%


Como retorno da execução, teremos a taxa de acerto de 76,34% e a taxa de palavras desconhecidas de 6.99%.

A seguir, descobriremos se há um maneira de tornarmos nosso algoritmo mais abrangente com alguma expansão de metodologia.

Após algumas reflexões sobre o corretor, poderemos chegar em algumas conclusões; uma delas diz respeito à variável palavra que é a palavra “lóigica”.

In [55]:
palavra = "lóigica"

Precisaremos corrigir este termo equivocado. Para isso, eliminaremos a letra “i” para obtermos o correto “lógica”. Nesta correção, realizamos uma única operação que deleta apenas um caractere.

Porém, às vezes poderemos ter um problema em que sejam necessárias mais operações para corrigirmos a palavra digitada equivocadamente.

Por exemplo, se escrevermos erroneamente “lóiigica”, precisaremos deletar o caractere “i” duas vezes, e o nosso corretor ainda não é capaz de fazer isso.

Logo, notaremos que o nosso corretor só corrige palavras que estão a uma única distância da correta. Então construiremos uma função que corrigirá as que estão a duas distâncias do resultado correto, e verificará qual será o retorno do novo corretor.

Para isso, criaremos um novo gerador de palavras chamado gerador_turbinado() com def, o qual receberá as palavras_geradas do próprio gerador_palavras como parâmetro.

Esta função retornará os novos vocábulos corrigidos a duas distâncias da palavra digitada equivocadamente, ou seja, novas_palavras.

Pegaremos cada uma das palavras geradas pelo gerador_palavras com for, aplicando novamente o gerador para palavra.

No caso da palavra “lóiigica”, o gerador_palavras conterá a palavra 'lóigica' como resultado. A partir daí, o gerador_turbinado() removerá o outro caractere “i” a mais, gerando a palavra correta "lógica".

Então, para cada palavra das palavras_geradas, chamaremos a função gerador_palavras() novamente e colocaremos tudo dentro de novas_palavras que será igual a uma lista.

Não aplicaremos .append() no gerador, pois seu retorno será uma lista. Logo, a concatenaremos com novas_palavras usando +=.

Com isso, retornaremos as novas palavras corrigidas a duas distâncias da digitada equivocadamente.

Para testarmos, chamaremos a nova variável temporária palavras_g ao final do bloco, chamando o gerador_turbinado() que receberá as palavras_geradas() com a palavra digitada equivocadamente “lóigica”, ou seja, a palavra.

Agora queremos saber se o termo “lógica” está de fato em palavras_g. Caso o retorno seja verdadeiro, o algoritmo de gerador de palavras turbinado corrigirá este tipo de erro.

In [56]:
palavra = "lóiigica"

def gerador_turbinado(palavras_geradas):
    novas_palavras = []
    for palavra in palavras_geradas:
        novas_palavras += gerador_palavras(palavra)
    return novas_palavras

palavras_g = gerador_turbinado(gerador_palavras(palavra))
"lógica" in palavras_g

True

Rodando o código, teremos o resultado True.

Portanto, nosso gerador_turbinado() está corrigindo palavras com até duas de distância da correta.

Para descobrirmos o tamanho desta nova lista, aplicaremos len() das palavras_g na célula seguinte.

In [57]:
len(palavras_g)

676760

Como retorno, veremos que gerou 676.760 variações de uma única palavra, ou seja teremos um volume extra de palavras geradas.

Sempre fazer a correção com esta grande quantidade de resultados talvez não faça sentido, pois demoraria muito.

A seguir, veremos uma maneira de eliminar as palavras que não existem dentro de palavras_geradas.

Anteriormente, vimos um gigantesco número de palavras geradas para uma única correção, tanto pelo gerador_palavras() quanto pelo gerador_turbinado().

Por conta disso, calcular a máxima probabilidade de todas esses possíveis resultados corretos se tornou inviável, então precisaremos reduzir este número de candidatos.

Construímos a função que chama o gerador_palavras() pelo corretor(), a qual gera as palavras e calcula a probabilidade de cada uma, retornando a possível correção.

Teremos que modificá-la; copiaremos seu bloco com "Ctrl + C" e colaremos em uma nova célula com "Ctrl + V".

Chamaremos este corretor de novo_corretor(), pois ainda utilizaremos o antigo mais adiante, e passaremos as palavras_geradas para o meu gerador_turbinado().

In [58]:
def novo_corretor(palavra):
    palavras_geradas = gerador_palavras(palavra)
    palavra_correta = max(palavras_geradas, key=probabilidade)
    return palavra_correta

Então, apenas chamaremos função do turbinado para gerarmos palavras apenas através deste. Estas serão armazenadas em palavras_turbinado sendo igual a gerador_turbinado() recebendo palavras_geradas.

Em seguida, combinaremos essas duas variáveis. Antes, teremos uma outra variável chamada de todas_palavras, a qual será o conjunto set() das palavras_geradas somado com as palavras_turbinado. Depois, deveremos calcular a probabilidade de todas as palavras.

Porém, nosso problema diz respeito à enorme quantidade de resultados, então precisaremos pegar somente os candidatos, e para isso criaremos a variável candidatos.

Esta será uma lista com a própria palavra que queremos corrigir, visto que se não conseguirmos gerar uma possível correção, retornaremos a palavra a ser corrigida. Nesta, adicionaremos as palavras que existem no nosso vocabulário, do contrário a probabilidade sempre será zero e o vocábulo nunca será escolhido.

Para verificarmos quais palavras da todas_palavras existem no meu vocabulário, colocaremos um for de palavra nesta variável. Se estiver no vocabulario com if, a adicionaremos à lista candidatos com .append().

Feito isso, reduziremos o número de possíveis termos a serem corrigidos. Retornaremos a palavra_correta, porém a probabilidade será calculada a partir dos candidatos, e não mais de palavras_geradas.

Para verificarmos a quantidade de candidatos, colocaremos um print() do tamanho len() da variável candidatos. Depois, chamaremos essa função passando a palavra “lóigica” com novo_corretor() recebendo palavra.

In [59]:
def novo_corretor(palavra):
    palavras_geradas = gerador_palavras(palavra)
    palavras_turbinado = gerador_turbinado(palavras_geradas)
    todas_palavras = set(palavras_geradas + palavras_turbinado)
    candidatos = [palavra]
    for palavra in todas_palavras:
        if palavra in vocabulario:
            candidatos.append(palavra)
    print(len(candidatos))
    palavra_correta = max(candidatos, key=probabilidade)
    return palavra_correta

novo_corretor(palavra)

2


'lógica'

Com isso, corrigimos a nossa palavra ao receber 'lógica', e calculamos a probabilidade máxima de candidatos, sendo apenas 2. O grande restante de resultados é insignificante para este novo corretor, pois não nos servirá para o que queremos.

Portanto, conseguimos otimizar muito o nosso resultado, de quase um milhão de palavras para apenas duas possíveis correções.

A seguir, trabalharemos com os dados de teste.

# Avaliando o resultado dos dois corretores

Neste passo, testaremos o novo_corretor() e o compararemos com o corretor antigo para verificarmos as qualidades e escolhermos o melhor.

Começaremos adaptando a função avaliador() copiando e colando seu bloco em uma nova célula. Depois, alteraremos a chamada para novo_corretor(), pois queremos também que esta realize a avaliação do novo corretor.

In [60]:
def avaliador(testes, vocabulario):
    numero_palavras = len(testes)
    acertou = 0
    desconhecida = 0
    for correta, errada in testes:
        palavra_corrigida = novo_corretor(errada)
        if palavra_corrigida == correta:
            acertou += 1
        else:
            desconhecida += (correta not in vocabulario)
    taxa_acerto = round(acertou*100/numero_palavras, 2)
    taxa_desconhecida = round(desconhecida*100/numero_palavras, 2)
    print(f"{taxa_acerto}% de {numero_palavras} palavras, desconhecidas {taxa_desconhecida}%")

vocabulario = set(lista_normalizada)
avaliador(lista_teste, vocabulario)

5
35
75
7
21
122
13
4
26
4
143
18
22
11
8
14
8
19
58
5
8
51
5
111
30
12
9
25
10
26
3
41
4
14
3
9
4
6
144
108
133
5
6
184
5
6
4
88
17
4
5
15
176
5
38
268
14
32
34
8
43
58
6
5
49
57
63
23
69
38
4
14
211
7
64
76
41
4
93
4
38
6
108
32
8
21
10
18
162
11
4
122
29
46
38
16
3
13
6
69
3
41
4
35
3
18
5
2
32
74
70
4
6
10
2
7
4
9
17
6
5
18
191
5
32
104
11
4
4
61
56
60
5
5
27
67
45
23
162
25
16
96
2
2
3
12
10
2
34
2
18
238
2
3
4
4
5
4
24
5
4
13
1
6
3
13
2
4
6
1
4
1
9
1
1
5
3
5
5
15
73
3
13
3
2
2
55.38% de 186 palavras, desconhecidas 6.45%


55.38% de 186 palavras, desconhecidas 6.45%

Antes disso, modificaremos a função novo_corretor() removendo o print() para que não imprima a quantidade de candidatos para toda a palavra dos dados de teste, pois essa informação não nos interessa mais.

In [66]:
def novo_corretor(palavra):
    palavras_geradas = gerador_palavras(palavra)
    palavras_turbinado = gerador_turbinado(palavras_geradas)
    todas_palavras = set(palavras_geradas + palavras_turbinado)
    candidatos = [palavra]
    for palavra in todas_palavras:
      if palavra in vocabulario:
        candidatos.append(palavra)
        palavra_correta = max(candidatos, key=probabilidade)
    return palavra_correta

novo_corretor(palavra)

'lógica'

Após a execução, o resultado será 55.91% de 186 palavras, sendo 6.45% a taxa de desconhecidas. Logo, tivemos uma redução das duas taxas.

O número de palavras desconhecidas teoricamente deveria ser um número fixo, mas este valor caiu.

Isso aconteceu porque colocamos o contador da desconhecida após o if, então só entramos nesse else caso a palavra_corrigida esteja incorreta. Porém, esta variável é fixa e independe se o corretor acertou ou não.

Precisaremos ter certa atenção, pois às vezes teremos variáveis de valor fixo e as tornamos de valor variável, fazendo com que esses resultados se alterem sem fazer algum sentido, como neste caso.

Para corrigirmos isso, removeremos o else para não cometermos este erro. Colocaremos o incremento desta variável antes do if mesmo, pois é independente deste. Se acertarmos ou errarmos a correção da palavra, o valor desconhecida será fixo.

In [67]:
def avaliador(testes, vocabulario):
    numero_palavras = len(testes)
    acertou = 0
    desconhecida = 0
    for correta, errada in testes:
      palavra_corrigida = novo_corretor(errada)
      desconhecida += (correta not in vocabulario)
      if palavra_corrigida == correta:
        acertou += 1
    taxa_acerto = round(acertou*100/numero_palavras, 2)
    taxa_desconhecida = round(desconhecida*100/numero_palavras, 2)
    print(f"{taxa_acerto}% de {numero_palavras} palavras, desconhecidas {taxa_desconhecida}%")

vocabulario = set(lista_normalizada)
avaliador(lista_teste, vocabulario)

UnboundLocalError: ignored

Com isso, teremos o resultado correto da minha taxa de desconhecida com 6.99%.

A seguir, verificaremos o valor de 55.91% da taxa de acertos.

Vimos que o novo_corretor() performou consideravelmente pior do que o nosso corretor anterior. Printaremos alguns resultados para analisarmos o porquê deste novo não estar indo tão bem quanto o outro.

Logo após o if, colocaremos um else para toda vez em que errarmos a palavra a ser corrigida. Imprimiremos a palavra errada e concatenaremos com um hífen para separarmos da palavra errada gerada pelo antigo corretor().

Em seguida, concatenaremos da mesma forma com a palavra corrigida pelo novo_corretor().

Imprimiremos a equivocada, a corrigida pelo meu antigo corretor e a corrigida pelo meu novo corretor para analisarmos os resultados.

In [ ]:
def avaliador(testes, vocabulario):
    numero_palavras = len(testes)
    acertou = 0
    desconhecida = 0
    for correta, errada in testes:
        palavra_corrigida = novo_corretor(errada)
        desconhecida += (correta not in vocabulario)
        if palavra_corrigida == correta:
            acertou += 1
        else:
            print(errada + "-" + corretor(errada) + "-" + palavra_corrigida)
    taxa_acerto = round(acertou*100/numero_palavras, 2)
    taxa_desconhecida = round(desconhecida*100/numero_palavras, 2)
    print(f"{taxa_acerto}% de {numero_palavras} palavras, desconhecidas {taxa_desconhecida}%")

vocabulario = set(lista_normalizada)
avaliador(lista_teste, vocabulario)

Com a execução, veremos uma lista em que a primeira palavra é a errada, a segunda é a gerada pelo corretor antigo, e a terceira pelo nosso novo.

Na primeira linha, teremos a palavra digitada equivocadamente esje, onde provavelmente a pessoa substituiu o “j” pelo “s”. Assim, gerou as palavras certas esse e se.

Na terceira linha do resultado, teremos a palavra dosa onde digitaram um “a” a mais, pois está ao lado da tecla "s".

Nosso corretor antigo fez a corrigiu para dos corretamente. Porém, o novo corretor corrige para do, o que está incorreto, pois novo_corretor gera ambos os resultados.

Porém, do aparece muito mais vezes do que a palavra dos. Como dependemos dessa máxima probabilidade, a função exibirá o termo que possui uma maior chance de estar correta, mesmo que a outra palavra também seja uma possibilidade correta.

Muitas vezes nossos erros de digitação estão apenas a uma distância da palavra equivocada, ou seja, é mais comum esquecermos de digitar ou trocarmos uma única letra.

Embora teoricamente o novo_corretor() parecesse ser melhor, está performando bem pior do que o outro. Portanto, é mais interessante continuarmos com o antigo.

Modificaremos o avaliador() novamente. Copiaremos seu bloco e o colaremos em uma nova célula, e ao invés de chamarmos o novo corretor, o deletaremos e chamaremos o antigo corretor(). O print() em else também não nos interessa mais, então poderemos apagá-lo.

In [69]:
def avaliador(testes, vocabulario):
    numero_palavras = len(testes)
    acertou = 0
    desconhecida = 0
    for correta, errada in testes:
        palavra_corrigida = corretor(errada)
        desconhecida += (correta not in vocabulario)
        if palavra_corrigida == correta:
            acertou += 1
    taxa_acerto = round(acertou*100/numero_palavras, 2)
    taxa_desconhecida = round(desconhecida*100/numero_palavras, 2)
    print(f"{taxa_acerto}% de {numero_palavras} palavras, desconhecidas {taxa_desconhecida}%")

vocabulario = set(lista_normalizada)
avaliador(lista_teste, vocabulario)

76.34% de 186 palavras, desconhecidas 6.99%


Rodando este código para verificarmos o resultado final do avaliador(), veremos nosso corretor tem 76.34% e desconhece aproximadamente 7% das palavras. Portanto, possui uma performance bem interessante.

Para finalizarmos, teremos uma nova célula para analisarmos seu funcionamento com outros exemplos de termos errados.

Imprimiremos o novo_corretor() e o antigo corretor para os nossos dados. Como exemplo, teremos a palavra sendo igual a “lgica”.

In [70]:
palavra = "lgica"
print(novo_corretor(palavra))
print(corretor(palavra))

fica
lógica


O novo corretor acaba corrigindo para fica, e o nosso corretor antigo e ideal para esses nossos dados fez a correção correta.

Já no caso da palavra "lóigica", os dois retornam a palavra correta.

Se eu colocarmos mais um “i” equivocadamente obtendo "lóiigica", veremos que nosso corretor definitivo não conseguirá corrigir esse tipo de erro.

Como retorno, veremos um alóiigica e a palavra correta lógica.

Poderemos testar mais e verificar que nosso algoritmo de corretor é eficiente e corrige muitas palavras.

Para melhorarmos ainda mais, poderemos procurar uma base de dados maior para fazermos esse treinamento e termos uma performance ainda melhor.

Finalizamos mais um curso da Plataforma Alura, e neste passo revisaremos o que realizamos em nossas aulas.

Começamos a discutir um pouco sobre o que é o Processamento de Linguagem Natural, e como como o ser humano e a máquina conseguem criar uma comunicação entre si, e percebemos não é feita de maneira direta.

O NLP é o intermediário, e possui diversas aplicações. Entre elas, temos o assistente pessoal, tradutor, os buscadores, e o corretor ortográfico, que é justamente o que criamos nesse curso.

Como exemplo, temos uma palavra digitada de maneira equivocada, geramos as possíveis palavras corretas, e então inferimos quais daquelas possíveis candidatas era justamente a correta.

Para as gerarmos, discutimos alguns algoritmos; o insere_letras(), onde nós inseríamos letras na palavra digitada equivocadamente, o deletando_caracteres(), o troca_letra() e o inverte_letra(), que é quando digitamos sem querer uma letra na frente da outra.

Esses algoritmos eram os responsáveis por gerar as palavras, mas não por realizar a inferência da palavra correta. Para isso, precisamos de uma base de dados, e neste caso foram os artigos do Blog da Alura compilados em artigos.txt.

Esse nosso arquivo é o nosso corpus textual com um número muito grande de caracteres, mas descobrimos também o número de palavras.

Separamos a string em pequenos tokens. Para isso, nós utilizamos a biblioteca Python específica para processamento de linguagem natural chamada nltk.

Depois, separamos dos tokens apenas os que eram palavras, e aí contamos o número de palavras que havia no nosso dataset., chegando a quase 400 mil, o que é um número bem interessante para o nosso treinamento.

Saber só o número de palavras não era suficiente, então contamos também os tipos de palavras. Após isso, iniciamos a construção do nosso algoritmo.

Começamos por insere_letras() e testamos esse nosso corretor apenas com esta função, e descobrimos que o resultado não era muito bom, pois corrigia apenas 1% dos dados, um número bem abaixo do que precisamos.

Implementamos os outros algoritmos e atingimos um resultado de quase 77%, já um número bem mais interessante.

Após atingirmos esse resultado, tentamos melhorar ainda mais o nosso corretor e criamos o gerador_turbinado(), o qual foi capaz de corrigir palavras como o “lóiigica” aqui.

Chegamos à conclusão que esse gerador turbinado não era muito bom na realidade, e o nosso novo corretor foi descartado, chegando à conclusão que o melhor resultado que nós atingimos foi realmente os quase 77% do primeiro corretor.

Portanto, ficamos com o antigo.

Se tiver qualquer dúvida, acesse nosso fórum que responderemos o mais rápido possível. Também convidamos a realizar todos os exercícios disponibilizados nesse curso.